### PR TITLE
feat: add convergence threshold for q law

### PR DESCRIPTION
**Pull request by vishwa2710**

### CODE REVIEW

**Review by apaletta3 — APPROVED** [review]
Took a brief look cause I was curious, looks nice!
It might be useful to create a Qlaw fixed constructor for our nominal case (LEO circular orbit station keeping or orbit raising) with the recommend weights and convergence tolerances pre-defined

File: include/OpenSpaceToolkit/Astrodynamics/GuidanceLaw/QLaw.hpp [hunk 1]
```diff
@@ -61,13 +63,14 @@ class QLaw : public GuidanceLaw
...
-            const Map<COE::Element, double>& anElementWeights,
+            const Map<COE::Element, Tuple<double, double>>& anElementWeightsMap,
             const Size& aMValue = 3,
             const Size& aNValue = 4,
             const Size& aRValue = 2,
+            const double& aBValue = 0.01,
             const Size& aKValue = 100,
-            const Length& minimumPeriapsisradius = Length::Kilometers(6578.0),
-            const double& aBValue = 0.01
+            const double& aPeriapsisWeight = 0.0,
```
Comment (Derollez) [inline include/OpenSpaceToolkit/Astrodynamics/GuidanceLaw/QLaw.hpp:72]: Alright, that actually answers of the question/suggestion I had to make this parameterizable ✅

File: src/OpenSpaceToolkit/Astrodynamics/GuidanceLaw/QLaw.cpp [hunk 2]
```diff
@@ -193,6 +178,21 @@ Vector3d QLaw::computeThrustDirection(const Vector6d& aCOEVector, const double&
...
+        (std::acos(std::cos((aCOEVector[3] - targetCOEVector_[3])))),
+        (std::acos(std::cos((aCOEVector[4] - targetCOEVector_[4])))),
+    };
+
+    // Within tolerance of all targeted elements. No need to thrust.
+    if ((parameters_.controlWeights_.array() * deltaCOE.array().abs() <= parameters_.convergenceThresholds_.array())
+            .all())
+    {
+        return {0.0, 0.0, 0.0};
+    }
```
Comment (Derollez) [inline src/OpenSpaceToolkit/Astrodynamics/GuidanceLaw/QLaw.cpp:194]: Does controlWeights ensure the delta are all scaled properly and end-up in the same range?

**Review by Derollez — APPROVED** [review]
Quick comment, LGTM otherwise!

File: src/OpenSpaceToolkit/Astrodynamics/GuidanceLaw/QLaw.cpp [hunk 3]
```diff
@@ -193,6 +178,21 @@ Vector3d QLaw::computeThrustDirection(const Vector6d& aCOEVector, const double&
...
+    const Vector5d deltaCOE = {
+        (aCOEVector[0] - targetCOEVector_[0]),
+        (aCOEVector[1] - targetCOEVector_[1]),
+        (aCOEVector[2] - targetCOEVector_[2]),
+        (std::acos(std::cos((aCOEVector[3] - targetCOEVector_[3])))),
+        (std::acos(std::cos((aCOEVector[4] - targetCOEVector_[4])))),
+    };
+
+    // Within tolerance of all targeted elements. No need to thrust.
+    if ((parameters_.controlWeights_.array() * deltaCOE.array().abs() <= parameters_.convergenceThresholds_.array())
```
Comment (Derollez) [inline src/OpenSpaceToolkit/Astrodynamics/GuidanceLaw/QLaw.cpp:190]: side: wondering if we could use some kind of mean metrics here? 🤔

File: src/OpenSpaceToolkit/Astrodynamics/GuidanceLaw/QLaw.cpp (hunk 2 above)
Comment (vishwa2710) [reply src/OpenSpaceToolkit/Astrodynamics/GuidanceLaw/QLaw.cpp:184]: Yeah the control weights ensure that we are ignoring the delta between the "free" variables.

**Pull request merged**